### ISSUE TITLE
`twisted.web.resource.IResource` doesn't define `server` attribute used by `Resource`

### ISSUE THREAD

**Issue by twm** [twisted/twisted#11717]
**Describe the incorrect behavior you saw**

The `twisted.web.resource.Resource.putChild()` implementation touches the `server` attribute of its `child` argument, which is nominally an `IResource`:

https://github.com/twisted/twisted/blob/4b62da085e6a7b8e4b1231e8cdf16e0d3439591d/src/twisted/web/resource.py#L207-L233

`IResource` doesn't define a `server` attribute, much less a mutable one.

This implies that an `IResource` is associated with a single `server`, which isn't mentioned in `IResource` either.

**Describe how to cause this behavior**

Type annotate the method like `def putChild(self, path: bytes, child: IResource) -> None`.

**Describe the correct behavior you'd like to see**

Options:

1. Define `IResource.server` and describe this behavior in `IResource.putChild()`
2. Annotate `Resource.putChild` as accepting `IResourceAsItExistsInTheRealWorld` (see also: the many attributes that `DummyResource` has to define in order to function).
3. Something more complicated in the name of backward compatibility.
4. Rewrite Twisted Web. Why not give it a try?

**Comment by adiroiban** [comment]
I don't think that rewriting  Twisted web is an option... I remember there was twisted.web2 and it failed... and by now, we are in the web3.0 age :)

I would say that for not option 1 is the first step.

For the future, we can create a new `twisted.web.resource.ModernResource` that should have a better URL traversing and support for 100-continue and receiving the content as a stream.
